src/GraphQL/Types/Schema.cs

```diff
@@ -8,6 +8,8 @@ namespace GraphQL.Types
 {
     public interface ISchema : IDisposable
     {
+        IServiceProvider Services { get; }
```

Comment:
It seems to me that you should not give access to this property through the interface. This encourages users to use the ServiceLocator anti-pattern in user-code (mostly in resolvers). I understand that the `IServiceProvider` is already in the scheme, but it is used for the internal needs of the scheme itself. User-code can resolve dependencies without a ServiceLocator using standard dependency injection in constructor - #1160.

Reply:
Yeah that was a stop-gap.  I like that pattern of the factory.  I'll remove this.  I'm still thinking on how to fix it without the user needing to do any special tricks.  As a note, this Schema First does not have the same issue as the GraphType First approach.  The resolver classes get resolved on each request.

Reply:
The factory creates something definite and is used by certain classes. If an object creates anything and is accessible from anywhere, then this is ServiceLocator.

What issue?

Reply:
The dependency scoping issue.

Reply:
I did not quite understand, is there any difference in the way a schema was created? The result is the same. In any case, it’s all about not introducing client code into the temptation to use ServiceLocator. I carefully read #648 and saw in given examples all those rakes that I myself went to. All these attempts at any cost simply drag the magic object into any point of the code in order to create anything through it. This puts the whole point o DI upside down - letting the container itself create dependencies and pass them on to the constructors.

By the way, we are now successfully applying the approach proposed in #1160 in production - no manual resolving, no ServiceLocator, no tight coupling with `HttpContext/IHttpContextAccessor` + also simultaneous support of multiple scope types in one application.

Reply:
> is there any difference in the way a schema was created

Yes - today, if you use the GraphType First approach then your GraphType's will automatically take upon the lifetime of the Schema.  That won't happen with Schema First because the field resolver type is always fetched from the container each time a field is executed.  This is because with Schema First your schema definition and your resolvers are two different classes.

> By the way, we are now successfully applying the approach

Awesome!